allennlp/models/bidirectional_lm.py

```diff
@@ -218,65 +212,61 @@ def forward(self,  # type: ignore
         # pylint: disable=arguments-differ
         mask = get_text_field_mask(source)
 
-        # We must have token_ids so that we can compute targets
-        token_ids = source.get("tokens")
-        if token_ids is None:
-            raise ConfigurationError("Your data must have a 'tokens': SingleIdTokenIndexer() "
-                                     "in order to use the BidirectionalLM")
-
-        # Use token_ids to compute targets
-        forward_targets = torch.zeros_like(token_ids)
-        backward_targets = torch.zeros_like(token_ids)
-        forward_targets[:, 0:-1] = token_ids[:, 1:]
-        backward_targets[:, 1:] = token_ids[:, 0:-1]
-
-        # shape (batch_size, timesteps + 2, embedding_size)
+        # shape (batch_size, timesteps, embedding_size)
         embeddings = self._text_field_embedder(source)
-
         contextual_embeddings = self._contextualizer(embeddings, mask)
 
-        # add dropout
-        contextual_embeddings = self._dropout(contextual_embeddings)
-
-        # compute softmax loss
-        forward_loss, backward_loss = self._compute_loss(contextual_embeddings,
-                                                         embeddings,
-                                                         forward_targets,
-                                                         backward_targets)
+        return_dict = {}
 
-        num_targets = torch.sum((forward_targets > 0).long())
-        if num_targets > 0:
-            average_loss = 0.5 * (forward_loss + backward_loss) / num_targets.float()
-        else:
-            average_loss = torch.tensor(0.0).to(forward_targets.device)  # pylint: disable=not-callable
-        # this is stored to compute perplexity if needed
-        self._last_average_loss[0] = average_loss.detach().item()
-
-        if num_targets > 0:
-            # loss is directly minimized
-            if self._loss_scale == 'n_samples':
-                scale_factor = num_targets.float()
+        # If we have target tokens, calculate the loss.
+        token_ids = source.get("tokens")
+        if token_ids is not None:
+            # Use token_ids to compute targets
+            forward_targets = torch.zeros_like(token_ids)
+            backward_targets = torch.zeros_like(token_ids)
+            forward_targets[:, 0:-1] = token_ids[:, 1:]
+            backward_targets[:, 1:] = token_ids[:, 0:-1]
+
+            # add dropout
+            contextual_embeddings = self._dropout(contextual_embeddings)
```

Comment:
I feel like I'd rather have this dropout up where it was, just so that when you're reading the model it's obvious it's taking place. (it's less obvious when it happens somewhere else.)

Reply:
Let's document in the docstring where dropout is applied to make it explicit.

Also, this will have different behavior for the output `lm_embeddings` key depending on whether token_ids is provided as input (with token_ids, dropout will be applied, but won't be applied if token_ids is not provided).  This is confusing.  Recommend creating a new tensor here to compute the loss, and always returning `lm_embeddings` that has not had dropout applied.